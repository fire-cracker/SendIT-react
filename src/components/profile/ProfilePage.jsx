import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import decodeJwt from 'jwt-decode';

import { getOrder } from '../../redux/actions/track';
import OrderEntries from './OrderEntries';
import Header from '../header';
import './style.css';

export class ProfilePage extends Component {
  state = {
    showNew: false,
    showPending: true,
    showDelivered: true,
    status: 'New'
  }

  componentDidMount() {
    const { token } = localStorage;
    const { userId } = decodeJwt(token);
    const { getOrder, history } = this.props;
    getOrder(userId)
      .then()
      .catch((error) => {
        const { response, response: { status } } = error;
        if (response && status === 401) {
          history.push('/');
          return toast.error('Your session has expired. You need to login');
        }
        toast.error('Unknown error');
      });
  }

  componentDidUpdate() {
    const { history, login: { isLoggedIn } } = this.props;
    if (isLoggedIn === false) {
      history.push('/');
    }
  }

  toggleDiv = (status, showingDiv) => {
    this.setState(() => ({
      showNew: true,
      showPending: true,
      showDelivered: true,
      [showingDiv]: false,
      status
    }));
  };

  render() {
    const { orders: { orders } } = this.props;
    const {
      showNew, showPending, showDelivered, status
    } = this.state;
    const deliveredOrders = orders.filter(order => order.orderStatus === 'Delivered');
    const deliveredOrdersLength = deliveredOrders.length;
    const unresolvedOrders = orders.filter(order => order.orderStatus === 'New' || 'Pending');
    const unresolvedOrdersLength = unresolvedOrders.length;
    const { token } = localStorage;
    const { userName } = decodeJwt(token);
    const orderEntries = orders.filter(order => order.orderStatus === status);

    return (
      <div className="profile__page">
        <div>
          <Header />
        </div>
        <div>
          <div className="createAshipment">
            <h1 id="user">
            Hi!  &nbsp;
              {userName}
            </h1>

            <form className="profileDetails" id="profileDetails">
              <div className="column">
                <div className="row1">
                  <h2> Delivered parcel delivery order</h2>
                  <a id="deliveredNumber">{deliveredOrdersLength}</a>
                </div>
                <div className="row2">
                  <h2>Pending parcel delivery order</h2>
                  <a id="pendingNumber">{unresolvedOrdersLength}</a>
                </div>
              </div>
            </form>


          </div>
          <div id="informationFields">
            <div className="profile__preview">
              <h1> Parcel Delivery Orders</h1>

              <div className="navTabs" id="navTabs">
                <a
                  className={showNew ? 'navLink' : 'navLink active'}
                  onClick={() => this.toggleDiv('New', 'showNew')}
                >
                NEW ORDERS
                </a>
                <a
                  className={showPending ? 'navLink' : 'navLink active'}
                  onClick={() => this.toggleDiv('Pending', 'showPending')}
                >
                PENDING ORDERS
                </a>
                <a
                  className={showDelivered ? 'navLink' : 'navLink active'}
                  onClick={() => this.toggleDiv('Delivered', 'showDelivered')}
                >
                DELIVERED ORDERS&nbsp;
                </a>
              </div>

              <div id="information">

                <div
                  id="shipmentOrder"
                  className="newPending-active"
                >
                  { orderEntries.length
                    ? (
                      <div className="order-table">
                        <table>
                          <thead>
                            <tr>
                              <th>Pickup Location</th>
                              <th>Destination</th>
                              <th>weight(kg)</th>
                              <th>Price</th>
                              <th>Date of Order</th>
                              <th>Status</th>
                              {showPending ? null : (<th>PresentLocation</th>)}
                            </tr>

                            <OrderEntries showPending={showPending} orders={orderEntries} />

                          </thead>
                        </table>
                      </div>
                    )
                    : (
                      <p className="no__order">No Order</p>
                    )
                  }

                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

ProfilePage.propTypes = {
  getOrder: PropTypes.func,
  orders: PropTypes.object,
  history: PropTypes.object,
  login: PropTypes.object
};
const mapStateToProps = ({ login, orders }) => ({ login, orders });

const mapDispatchToProps = {
  getOrder
};

export default connect(mapStateToProps, mapDispatchToProps)(ProfilePage);
