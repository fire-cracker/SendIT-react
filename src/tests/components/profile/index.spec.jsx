import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import decodeJwt from 'jwt-decode';
import { toast } from 'react-toastify';

import { ProfilePage } from '../../../components/profile/ProfilePage';
import { orders } from '../../mock/orders';

jest.mock('../../../utils/axiosConfig');
jest.mock('react-toastify');
jest.mock('../../../redux/actions/createOrder');
jest.mock('jwt-decode');

describe('Tests for the Profile Page', () => {
  const props = {
    getOrder: jest.fn().mockResolvedValue({}),
    orders: { orders },
    login: {
      isLoggedIn: true
    }
  };
  let wrapper;

  toast.error = jest.fn();

  decodeJwt.mockImplementation(() => ({
    userId: 'chinchin'
  }));

  beforeEach(() => {
    wrapper = mount(<MemoryRouter><ProfilePage {...props} /></MemoryRouter>);
  });

  it('Should render the Profile page', () => {
    expect(wrapper).toMatchSnapshot();
    wrapper.find('.navLink').at(0).simulate('click');
    wrapper.find('.navLink').at(1).simulate('click');
    wrapper.find('.navLink').at(2).simulate('click');
  });

  it('should throw error 401 if request is incorrect', () => {
    const props = {
      getOrder: jest.fn().mockImplementation(() => Promise.reject({
        response: {
          status: 401
        }
      })),
      orders: { orders }
    };
    wrapper = mount(<MemoryRouter><ProfilePage {...props} /></MemoryRouter>);
  });

  it('should throw error if request is not sucessful', () => {
    const props = {
      getOrder: jest.fn().mockImplementation(() => Promise.reject({
        response: {
          status: 500
        }
      })),
      orders: { orders }
    };
    wrapper = mount(<MemoryRouter><ProfilePage {...props} /></MemoryRouter>);
  });
});
