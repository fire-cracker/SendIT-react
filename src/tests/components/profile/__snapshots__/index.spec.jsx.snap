// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the Profile Page Should render the Profile page 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "frj8te",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "frj8te",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ProfilePage
      getOrder={
        [MockFunction] {
          "calls": Array [
            Array [
              "chinchin",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Promise {},
            },
          ],
        }
      }
      login={
        Object {
          "isLoggedIn": true,
        }
      }
      orders={
        Object {
          "orders": Array [
            Object {
              "fromAddress": "Oyedeji Haven, Jagun Old Ife Road Ibadan",
              "toAddress": "Oyedeji Haven, Jagun Old Ife Road Ibadan",
              "weight": "2",
            },
            Object {
              "fromAddress": "Oyedeji Haven, Jagun Old Ife Road Ibadan",
              "toAddress": "Oyedeji Haven, Jagun Old Ife Road Ibadan",
              "weight": "2",
            },
          ],
        }
      }
    >
      <div
        className="profile__page"
      >
        <div>
          <Header>
            <div
              className="header"
              id="header"
            >
              <ToastContainer
                autoClose={5000}
                bodyClassName={null}
                className={null}
                closeButton={
                  <CloseButton
                    ariaLabel="close"
                  />
                }
                closeOnClick={true}
                draggable={true}
                draggablePercent={80}
                hideProgressBar={false}
                newestOnTop={false}
                pauseOnFocusLoss={true}
                pauseOnHover={true}
                position="top-right"
                progressClassName={null}
                progressStyle={null}
                rtl={false}
                style={null}
                toastClassName={null}
                transition={[MockFunction]}
              />
              <header
                className="navBar"
                id="navBar"
              >
                <div
                  id="headerContent"
                >
                  <div
                    className="navigation"
                    id="navigation"
                  >
                    <Link
                      id="logout"
                      replace={false}
                      to="/"
                    >
                      <a
                        href="/"
                        id="logout"
                        onClick={[Function]}
                      >
                        Logout
                      </a>
                    </Link>
                    <Link
                      id="profile"
                      replace={false}
                      to="/profile"
                    >
                      <a
                        href="/profile"
                        id="profile"
                        onClick={[Function]}
                      >
                        Profile
                      </a>
                    </Link>
                    <Link
                      id="active"
                      replace={false}
                      to="/createOrder"
                    >
                      <a
                        href="/createOrder"
                        id="active"
                        onClick={[Function]}
                      >
                        Create Order
                      </a>
                    </Link>
                    <Link
                      id="track"
                      replace={false}
                      to="/track"
                    >
                      <a
                        href="/track"
                        id="track"
                        onClick={[Function]}
                      >
                        Track
                      </a>
                    </Link>
                    <a
                      className="icon"
                      id="icon"
                      onClick={[Function]}
                    >
                      <i
                        className="fa fa-bars"
                      />
                    </a>
                  </div>
                </div>
                <div
                  className="blink_caption"
                  href="home.html"
                >
                  <div
                    className="blink"
                    id="blin"
                  >
                    Just SendIT!!!
                  </div>
                </div>
              </header>
            </div>
          </Header>
        </div>
        <div>
          <div
            className="createAshipment"
          >
            <h1
              id="user"
            >
              Hi!   
            </h1>
            <form
              className="profileDetails"
              id="profileDetails"
            >
              <div
                className="column"
              >
                <div
                  className="row1"
                >
                  <h2>
                     Delivered parcel delivery order
                  </h2>
                  <a
                    id="deliveredNumber"
                  >
                    0
                  </a>
                </div>
                <div
                  className="row2"
                >
                  <h2>
                    Pending parcel delivery order
                  </h2>
                  <a
                    id="pendingNumber"
                  >
                    2
                  </a>
                </div>
              </div>
            </form>
          </div>
          <div
            id="informationFields"
          >
            <div
              className="profile__preview"
            >
              <h1>
                 Parcel Delivery Orders
              </h1>
              <div
                className="navTabs"
                id="navTabs"
              >
                <a
                  className="navLink active"
                  onClick={[Function]}
                >
                  NEW ORDERS
                </a>
                <a
                  className="navLink"
                  onClick={[Function]}
                >
                  PENDING ORDERS
                </a>
                <a
                  className="navLink"
                  onClick={[Function]}
                >
                  DELIVERED ORDERS 
                </a>
              </div>
              <div
                id="information"
              >
                <div
                  className="newPending-active"
                  id="shipmentOrder"
                >
                  <p
                    className="no__order"
                  >
                    No Order
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </ProfilePage>
  </Router>
</MemoryRouter>
`;
